File: geomajas/face/geomajas-face-dojo/geomajas-dojo-client/src/main/resources/geomajas/map/snapping/SnappingAlgorithm.js
/*
 * This file is part of Geomajas, a component framework for building
 * rich Internet applications (RIA) with sophisticated capabilities for the
 * display, analysis and management of geographic information.
 * It is a building block that allows developers to add maps
 * and other geographic data capabilities to their web applications.
 *
 * Copyright 2008-2010 Geosparc, http://www.geosparc.com, Belgium
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

dojo.provide("geomajas.map.snapping.SnappingAlgorithm");
dojo.declare("SnappingAlgorithm", null, {

	/**
	 * @fileoverview General abstraction for snapping algorithms.
	 * @class This abstract class needs to be extened by actual implementations
	 * of a snapping algorithm.
	 * @author Pieter De Graef
	 * 
	 * @constructor
	 * @param geometries Array of geometries on which to apply snapping.
	 */
	constructor : function (geometries) {
		this.geometries = geometries;
		this.minDistance = Number.MAX_VALUE;
	},

	/**
	 * The simplest of all snapping algorithms: Don't snap at all
 	 * @ret				  Returns closest snappoint if snapping applies, else 
	 *					  returns null.
	 */
	calcSnapPoint : function (coordinate, minDistance) {
		return null;
	},

	/**
	 * Getter for the minimum distance.
	 * @return Returns the current minimum distance.
	 */
	getMinDistance : function () {
		return this.minDistance;
	},

	/**
	 * @private
	 * This function retrieves the coordinates of a given geometry as a 2D
	 * array.
	 */
	_getAllCoordinates : function (geometry) {
		if (geometry instanceof MultiPolygon) {
			var coords = [];
			for (var i=0; i<geometry.getNumGeometries(); i++) {
				var polygon = geometry.getGeometryN(i);
				for (var j=0; j<polygon.getNumGeometries(); j++) { 
					/* j: 0 .. number of rings - 1 (interior + exterior) */
					coords.push (polygon.getGeometryN(j).getCoordinates());
				}
			}
			return coords;
		} else if (geometry instanceof Polygon) {
			var coords = [];
			for (var i=0; i<geometry.getNumGeometries(); i++) {
				coords[i] = geometry.getGeometryN(i).getCoordinates();
			}
			return coords;
		} else if (geometry instanceof MultiLineString) {
			var coords = [];
			for (var i=0; i<geometry.getNumGeometries(); i++) {
				coords[i] = geometry.getGeometryN(i).getCoordinates();
			}
			return coords;
		} else if (geometry instanceof LineString) {
			return [geometry.getCoordinates()];
		} else if (geometry instanceof Point) {
			return [geometry.getCoordinates()];
		}
	}
});

File: face/geomajas-face-gwt/client/src/main/java/org/geomajas/gwt/client/spatial/snapping/SnappingAlgorithm.java
/*
 * This file is part of Geomajas, a component framework for building
 * rich Internet applications (RIA) with sophisticated capabilities for the
 * display, analysis and management of geographic information.
 * It is a building block that allows developers to add maps
 * and other geographic data capabilities to their web applications.
 *
 * Copyright 2008-2010 Geosparc, http://www.geosparc.com, Belgium
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.geomajas.gwt.client.spatial.snapping;

import java.util.List;

import org.geomajas.geometry.Coordinate;
import org.geomajas.gwt.client.spatial.geometry.Geometry;

/**
 * <p>
 * General definition of a snapping algorithm. All snapping algorithms get their target geometries to snap to at
 * construction. It is recommended to set up the geometries at this time (if they need sorting or whatever). Later on
 * the "getSnappingPoint" method gets actual snapped coordinates for the given coordinate. Together with a snapped
 * coordinate, it is also possible to ask for the eventual snapping distance through the getMinimumDistance method.
 * </p>
 *
 * @author Pieter De Graef
 */
public abstract class SnappingAlgorithm {

	/**
	 * The minimum distance at which a coordinate was last snapped.
	 */
	protected double minimumDistance;

	// -------------------------------------------------------------------------
	// Constructor:
	// -------------------------------------------------------------------------

	/**
	 * The only constructor. Always requires a list of geometries.
	 */
	protected SnappingAlgorithm(List<Geometry> geometries) {
	}

	// -------------------------------------------------------------------------
	// Public methods:
	// -------------------------------------------------------------------------

	/**
	 * Calculates a snapping point from the given coordinate.
	 *
	 * @param original
	 *            The original and unsnapped coordinate.
	 * @param threshold
	 *            A threshold value that needs to be beaten in order to snap. Only if the distance between the original
	 *            and the candidate coordinate is smaller then this threshold, can the candidate coordinate be a snapped
	 *            coordinate.
	 * @return Returns the eventual snapped coordinate, or null if no snapping occurred.
	 */
	abstract Coordinate getSnappingPoint(Coordinate original, double threshold);

	/**
	 * Get the distance between the original and lastly snapped point. If the "getSnappingPoint" returned "null", then
	 * this will be Double.MAX_VALUE. This method will only contain a useful value, if a snapping point was actually
	 * found.
	 */
	public double getMinimumDistance() {
		return minimumDistance;
	}
}
